File: Leetcode_Everyday/2022/11/17_number-of-matching-subsequences.go
package main

import "fmt"

func upperBound(target int, nums []int) int {
	if len(nums) == 0 {
		return -1
	}
	left, right := 0, len(nums)-1
	for left < right {
		middle := (left + right) / 2
		if nums[middle] <= target {
			left = middle + 1
		} else {
			right = middle
		}
	}
	if nums[left] <= target {
		return -1
	}
	return left
}

func numMatchingSubseq(s string, words []string) int {
	var pos [26][]int
	for i, v := range s {
		pos[v-'a'] = append(pos[v-'a'], i)
	}
	cnt := 0
	for _, word := range words {
		matched := true
		cur := -1
		for _, c := range word {
			idx := upperBound(cur, pos[c-'a'])
			if idx == -1 {
				matched = false
				break
			}
			cur = pos[c-'a'][idx]
		}
		if matched {
			cnt++
		}
	}
	return cnt
}

func main() {
	s := "rwpddkvbnnuglnagtvamxkqtwhqgwbqgfbvgkwyuqkdwhzudsxvjubjgloeofnpjqlkdsqvruvabjrikfwronbrdyyjnakstqjac"
	words := []string{"lnagtva", "kvbnnuglnagtvamxkqtwhqgwbqgfbvgkwyuqkdwhzudsxvju", "rwpddkvbnnugln", "gloeofnpjqlkdsqvruvabjrikfwronbrdyyj", "vbgeinupkvgmgxeaaiuiyojmoqkahwvbpwugdainxciedbdkos", "mspuhbykmmumtveoighlcgpcapzczomshiblnvhjzqjlfkpina", "rgmliajkiknongrofpugfgajedxicdhxinzjakwnifvxwlokip", "fhepktaipapyrbylskxddypwmuuxyoivcewzrdwwlrlhqwzikq", "qatithxifaaiwyszlkgoljzkkweqkjjzvymedvclfxwcezqebx"}
	fmt.Println(numMatchingSubseq(s, words))
}

File: Leetcode_Everyday/2022/11/17_number-of-matching-subsequences.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words)
    {
        vector<int> m[26];
        for (int i = 0; i < s.size(); i++)
            m[s[i] - 'a'].push_back(i);
        int count = 0;
        for (string word : words)
        {
            bool match = true;
            int idx = -1;
            for (char c : word)
            {
                auto iter = upper_bound(m[c - 'a'].begin(), m[c - 'a'].end(), idx);
                if (iter == m[c - 'a'].end())
                {
                    match = false;
                    break;
                }
                idx = *iter;
            }
            if (match)
                count++;
        }
        return count;
        // vector<int> flag(26, -1);
        // vector<vector<int>> m(s.size(), vector<int>(26, -1));
        // for (int i = 0; i < s.size(); i++)
        // {
        //     if (flag[s[i] - 'a'] == -1)
        //         flag[s[i] - 'a'] = i;
        //     for (int j = i + 1; j < s.size(); j++)
        //     {
        //         if (m[i][s[j] - 'a'] == -1)
        //             m[i][s[j] - 'a'] = j;
        //     }
        // }
        // int count = 0;
        // for (string word : words)
        // {
        //     int idx = flag[word[0] - 'a'];
        //     if (idx == -1)
        //         continue;
        //     bool match = true;
        //     for (int i = 1; i < word.size(); i++)
        //     {
        //         if (m[idx][word[i] - 'a'] < idx)
        //         {
        //             match = false;
        //             break;
        //         }
        //         idx = m[idx][word[i] - 'a'];
        //     }
        //     if (match)
        //         count++;
        // }
        // return count;
    }
};
